src/state.rs

```diff
@@ -42,24 +45,26 @@ pub enum StateError {
 /// **Generics**:
 /// - O: The output timer used for keeping track of keep-alive ping-pongs. Must
 ///   implement the [`embedded_hal::timer::CountDown`] trait
-pub struct MqttState<P> {
+pub struct MqttState<P, T> {
     /// Connection status
     pub connection_status: MqttConnectionStatus,
     /// Status of last ping
     pub await_pingresp: bool,
     /// Packet id of the last outgoing packet
     pub last_pid: Pid,
     /// Outgoing QoS 1, 2 publishes which aren't acked yet
-    pub outgoing_pub: FnvIndexMap<u16, PublishRequest<P>, consts::U4>,
+    pub(super) outgoing_pub: FnvIndexMap<u16, Inflight<P, T>, consts::U4>,
```

Comment:
I think we could, with great improvements, change the `max_inflight` property of the config, to a generic `ArrayLength`, and then use it here? I don't see many cases where one wants to dynamically alter `max_inflight` after compile time?

Reply:
It leads to a signature below. And notablly the length should be `PowerOfTwo`. 
```rust
pub struct MqttState<P, T, L>
where
L: ArrayLength<Bucket<u16, Inflight<P, T>>> + ArrayLength<Option<Pos>> + PowerOfTwo
{
    pub(super) outgoing_pub: FnvIndexMap<u16, Inflight<P, T>, L>,
..
}
```
How about find a sensible default instead, say 8? Paho, for example, defaults to 10 or 1 (when a connection is reliable).

Reply:
Ugh, thats an ugly trait bound signature :( 

> How about find a sensible default instead, say 8? Paho, for example, defaults to 10 or 1 (when a connection is reliable).

I would have liked this, but as soon as we start saving the actual packet in there it escalates quickly. Currently our packet size in the factbird project is 1024 bytes, if we say a sensible default of 8, thats 8192 bytes of stack, locked in just "potential" inflights that will probably never be used. (I think something like 2, 3 or max 4 inflights makes sense for us, but that might be very low if eg. your payload is max 128 bytes)? We could just set it for our usage (3 for example) and then take the extended discussion when/if some other user needs it to be higher/bigger?

Reply:
Right. For now it's safe to leave it `consts::U4`. I'll be very happy to participate in the future discussion:)